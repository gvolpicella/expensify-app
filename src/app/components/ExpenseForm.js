import React from 'react';
import moment from 'moment';
import { SingleDatePicker } from 'react-dates';

const now = moment();

class ExpenseForm extends React.Component {
    constructor(props) {
        super(props);
        this.state = {
            description: props.expense ? props.expense.description : '',
            note: props.expense ? props.expense.note : '',
            amount: props.expense ? (props.expense.amount / 100).toString() : '',
            createdAt: props.expense ? moment(props.expense.createdAt) : moment(),
            calendarFocused: false,
            error: '',
            buttonText: props.expense ? 'Save expense'  : 'Add expense'
        };
    };   
    onDescriptionChange = (e) => {
        const description = e.target.value;
        this.setState(() => ({ description }));
    };
    onNoteChange = (e) => {
        const note = e.target.value;
        this.setState(() => ({ note }));
    };    
    onAmountChange = (e) => {
        const amount = e.target.value;

        if (!amount || amount.match(/^\d{1,}(\.\d{0,2})?$/)) {
            this.setState(() => ({ amount }));
        }  
    };
    onDateChange = (createdAt) => {
        if (createdAt) {
            this.setState(() => ({ createdAt }));
        }
    };    
    onFocusChange = ({ focused }) => {
        this.setState(() => ({ calendarFocused : focused }));
    };   
    onSubmit = (e) => {
        e.preventDefault();

        if (!this.state.description || !this.state.amount) {
            // set error state
            this.setState(() => ({ error: 'Please enter a description and amount.' }));
        }

        if (this.state.description && this.state.amount) {
            // clear error
            this.setState(() => ({ error: '' }));
            // wiring up the parent component with the state
            this.props.onSubmit({
                description : this.state.description,
                amount: parseFloat(this.state.amount, 10) * 100,
                createdAt: this.state.createdAt.valueOf(),
                note: this.state.note
            });
        }
    };
    render() {
        return (
            <div>
                { this.state.error && <p className="error">{ this.state.error }</p>}
                <form onSubmit={this.onSubmit}>
                    <label htmlFor="description">Description</label>
                    <input 
                        name="description" 
                        type="text"
                        placeholder="Description"
                        autoFocus 
                        value={this.state.description}
                        onChange={this.onDescriptionChange}
                    /><br/><br/>
                    <label htmlFor="amount">Amount</label>
                    <input 
                        name="amount" 
                        type="text"
                        placeholder="Amount" 
                        value={this.state.amount}
                        onChange={this.onAmountChange}
                    /><br/><br/>
                    <SingleDatePicker
                        date={this.state.createdAt}
                        onDateChange={this.onDateChange}
                        focused={this.state.calendarFocused}
                        onFocusChange={this.onFocusChange}
                        numberOfMonths={1}
                        isOutsideRange={ () => {false} }
                        displayFormat="DD MMM YYYY"
                    /><br/><br/>
                    <label htmlFor="notes">Notes</label>
                    <textarea 
                        name="note" 
                        placeholder="Enter a note for your expenses" 
                        value={this.state.note}
                        onChange={this.onNoteChange}
                    /><br/><br/>
                    <button type="submit">{this.state.buttonText}</button>
                </form>
            </div>
        )
    }
}

export default ExpenseForm;