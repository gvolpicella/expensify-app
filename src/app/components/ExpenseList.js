// Presentational component pattern
// stateless functional component
import React from 'react';
import { connect } from 'react-redux';
import ExpenseListItem from './ExpenseListItem';
import selectExpenses from '../selectors/expenses';

export const ExpenseList = (props) => (
    <div>
        <div className="expenses">
        {
            props.expenses.length === 0 ? (
                <p>No expenses</p>
            )
        : (
            props.expenses.map((expense, index) => (
                <ExpenseListItem key={expense.id} {...expense} />
            ))
        )
        }                
        </div>
    </div>
);

// maps the store to props
const mapStateToProps = (state) => {
    return {
        expenses: selectExpenses(state.expenses, state.filters)
    };
};

// Higher order component (HOC) for our expenses
// generates ExpenseList components connected to Store
export default connect(mapStateToProps)(ExpenseList);
