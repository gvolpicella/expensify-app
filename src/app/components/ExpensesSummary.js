import React from 'react';
import { connect } from 'react-redux';
import selectExpenses from '../selectors/expenses';
import { getExpensesTotal, getExpensesCount } from '../selectors/expenses-total';
import numeral from 'numeral';

export const ExpensesSummary = ({ expenseCount, expensesTotal }) => {
  const expenseWord = expenseCount === 1 ? 'expense' : 'expenses';
  const formattedExpensesTotal = numeral(expensesTotal / 100).format('$0,0.00');
  return (
    <div className="expenses-summary">
      <h2>Viewing {expenseCount} {expenseWord} totalling {formattedExpensesTotal}</h2>
    </div>
  );
};

const mapStateToProps = (state) => {
    const visibleExpenses = selectExpenses(state.expenses, state.filters);
    return {
        expenseCount: getExpensesCount(visibleExpenses),
        expensesTotal: getExpensesTotal(visibleExpenses)
    };
};

export default connect(mapStateToProps)(ExpensesSummary);
