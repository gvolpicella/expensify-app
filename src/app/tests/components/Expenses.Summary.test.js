import React from 'react';
import { configure, shallow } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import { ExpensesSummary } from '../../components/ExpensesSummary';
import expenses from '../fixtures/expenses';

configure({ adapter: new Adapter() });

test('should render ExpensesSummary expenses totals', () => {
  const wrapper = shallow(<ExpensesSummary expenseCount={1} expensesTotal={235} />);
  expect(wrapper).toMatchSnapshot();
});

test('should render ExpenseList with empty message', () => {
  const wrapper = shallow(<ExpensesSummary expenseCount={0} expensesTotal={0} />);
  expect(wrapper).toMatchSnapshot();
});