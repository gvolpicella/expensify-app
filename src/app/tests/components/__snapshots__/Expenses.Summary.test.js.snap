// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render ExpenseList with empty message 1`] = `
<div
  className="expenses-summary"
>
  <h2>
    Viewing 
    0
     
    expenses
     totalling 
    $0.00
  </h2>
</div>
`;

exports[`should render ExpensesSummary expenses totals 1`] = `
<div
  className="expenses-summary"
>
  <h2>
    Viewing 
    1
     
    expense
     totalling 
    $2.35
  </h2>
</div>
`;
